test/stmt/errors.swift

```diff
@@ -165,3 +165,26 @@ func thirteen() {
     }
   }
 }
+
+// SR 6400
+
+enum SR_6400_E: Error {
+  case castError
+}
+
+struct SR_6400_S_1 {}
+struct SR_6400_S_2: Error {}
+
+func sr_6400() throws {
+  do {
+    throw SR_6400_E.castError
+  } catch is SR_6400_S_1 { // expected-error {{type 'SR_6400_S_1' does not conform to protocol 'Error'}} expected-warning {{cast from 'Error' to unrelated type 'SR_6400_S_1' always fails}}
```

Comment:
:-/ Having two diagnostics seems redundant to me. I like just leaving the cast diagnostic there (the warning) because that'll take any wacky special cases into account like bridging or optional unwrapping. (I don't think those'll affect us here in practice, but it still doesn't seem worth having a separate implementation to check.)

Reply:
(At the time Timothy originally filed this, I don't know if we emitted the cast diagnostic.)

Reply:
@jrose-apple Seems like we didn't emit _any_ diagnostic for this case (I tested on 5.1, 5.0, 4.2 & 4.1). I added the protocol conformance diagnostic so the user would know what's wrong here - but if you think it's better to just have the cast diagnostic emitted, then I can remove the protocol one.

Reply:
Hm. I was thinking of something like…

```swift
do {
  throw SR_6400_E.castError as NSError
} catch let opaqueError as NSObject {
  print(opaqueError)
}
```

…which I expected to succeed because NSError is a subclass of NSObject. (Without the coercion-in-advance it's *probably* okay that that fails; it fails in a `switch` too.) But it looks like we don't actually do that in practice.

@DougGregor, would you expect that to catch the error or not? If `catch` really is identical to `case` it should succeed, but that's probably less useful than the current behavior.

Reply:
Hmm that doesn't get caught, but it's because `SR_6400_E` conforms to `Error`, so you'll have to cast it to `Error` in the `catch` statement for it to succeed.

Reply:
That's not how bridging usually works.

Reply:
Oh, yeah, in the above example, `opaqueError` doesn't end up being `NSError` (or `_SwiftNativeNSError`) in the catch statement so the cast to `NSObject` fails. Is that a bug i.e bridges error not being propagated to the catch statement?

Reply:
@jrose-apple I have removed the protocol conformance diagnostic, so now you'll just get the cast diagnostic now.